### PR TITLE
Support custom attributes

### LINKED ISSUE #45: name of email attribute in SAML callback response isn't expected by saml.rb
https://github.com/PracticallyGreen/omniauth-saml/blob/master/lib/omniauth/strategies/saml.rb#L93

It appears that saml.rb expects the email attribute to have a name of "email" or "mail".  The SAML IDP we are authenticating against returns "emailAddress".  I don't know if that is non-standard but we can't do anything about it.

I monkey patched the class to get this working in our environment.

Would you be open to adding optional configuration parameters to map the attributes to the values expected in the info block?

```ruby
option :email_attribute_name, :email
option :first_name_attribute_name, :first_name
option :last_name_attribute_name, :last_name
option :name_attribute_name, :name
...
info do
        {
          :name  => @attributes[options[:name_attribute_name]],
          :email => @attributes[options[:email_attribute_name]],
          :first_name => @attributes[options[:first_name_attribute_name]],
          :last_name => @attributes[options[:last_name_attribute_name]]
        }
end
```
The sample above may break some existing installs because currently the SAML runs through a few potential names for a few of these attributes.  So to replicate that behavior we could do the following:

```ruby
option :email_attribute_name, :email
option :first_name_attribute_name, :first_name
option :last_name_attribute_name, :last_name
option :name_attribute_name, :name
...
info do
        {
          :name  => @attributes[options[:name_attribute_name]],
          :email => @attributes[options[:email_attribute_name]] || @attributes[:mail],
          :first_name => @attributes[options[:first_name_attribute_name]]  || @attributes[:firstname] || @attributes[:firstName],
          :last_name => @attributes[options[:last_name_attribute_name]] || @attributes[:lastname] || @attributes[:lastName]
        }
end
```
I will work on the PR if you are receptive to this approach.

### PR DESCRIPTION
**Pull request by md5**
This PR replaces #51. Here is @OsQu's description from that PR:

> Attribute statements might not always have a Name="firstName" attribute, and modifying the idp to support can be an exhausting task (e.g. ADFS).
> 
> Now one can give attribute_statements option, which contains a mapping from statements to an info hash. For example the default mapping is following:

> ```
>   option :attribute_statements, {
    name: ["name"],
    email: ["email", "mail"],
    first_name: ["first_name", "firstname", "firstName"],
    last_name: ["last_name", "lastname", "lastName"]
  }
> ```
> This we can support whatever weird configurations the other end might have, without hard-coding them directly to the source.

Closes #45, #49, #51

### CODE REVIEW

**Comment by bufferoverflow** [comment]
:+1:

**Comment by OsQu** [comment]
:+1:!

**Comment by supernova32** [comment]
@md5 Great! thanks for resubmitting. Could we add a hint about this new option to the README as well? I think it should be well documented that it is possible to assign custom attributes and that changes to the idP server are not needed.

**Comment by md5** [comment]
Since this is from @ilikepi's branch, any additions would need to happen there unless we open yet another PR. Perhaps @ilikepi has some time to make those addition or someone can submit a PR to his branch.

**Comment by ilikepi** [comment]
I added a brief entry to the README.  Feedback welcome...I tried to be concise even though this particular option is a bit difficult to understand without some minimal amount of background knowledge.

**Comment by md5** [comment]
@ilikepi Looks good to me! :+1:

File: lib/omniauth/strategies/saml.rb [hunk 1]
```diff
@@ -103,15 +109,23 @@ def other_phase
...
       info do
-        {
-          :name  => @attributes[:name],
-          :email => @attributes[:email] || @attributes[:mail],
-          :first_name => @attributes[:first_name] || @attributes[:firstname] || @attributes[:firstName],
-          :last_name => @attributes[:last_name] || @attributes[:lastname] || @attributes[:lastName]
-        }
+        found_attributes = options.attribute_statements.map do |key, values|
+          attribute = find_attribute_by(values)
+          [key, attribute]
```
Comment (md5) [inline lib/omniauth/strategies/saml.rb]: Does it make sense to have `[key, nil]` here if no mapped attribute is found? I guess that's how the old code worked, so it's probably fine.

**Comment by ilikepi** [comment]
Updated the entry to avoid repeated use of the word "name".  Ugh...

File: lib/omniauth/strategies/saml.rb (hunk 1 above)
Comment (ilikepi) [inline lib/omniauth/strategies/saml.rb]: Hmm...the other option would be to treat it as a validation failure and raise an exception.  That seems overly aggressive, but admittedly I have very little experience with environments in which SAML is used.  There may very well be some environments in which it makes sense to treat a response with a "missing" Attribute as invalid.

**Comment by md5** [comment]
@ilikepi It might be good to give an example with `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress` or another URI-named attribute instead of just a plain name.

File: lib/omniauth/strategies/saml.rb (hunk 1 above)
Comment (md5) [inline lib/omniauth/strategies/saml.rb]: I'd still like to see this scheme extended to support multi-valued attributes as suggested in #36, so perhaps a similar approach could be used to add a `:required => true` flag.

Comment (ilikepi) [inline lib/omniauth/strategies/saml.rb]: Yeah, I wonder if it makes sense to address that in a separate PR.  It's starting to sound like it would be worth considering something more robust than a simple Hash for handling the various mapping scenarios...

Comment (md5) [inline lib/omniauth/strategies/saml.rb]: I think it should definitely be a separate PR.

Comment (supernova32) [inline lib/omniauth/strategies/saml.rb]: I agree :smiley:

**Comment by supernova32** [comment]
:+1: I think this is good to merge, since the suggested changes are going to be addressed in a different PR.
